apps/s_cb.c

```diff
@@ -1355,3 +1356,41 @@ void ssl_ctx_security_debug(SSL_CTX *ctx, int verbose)
     SSL_CTX_set_security_callback(ctx, security_callback_debug);
     SSL_CTX_set0_security_ex_data(ctx, &sdb);
 }
+
+static void keylog_callback(const SSL *ssl, const char *line)
+{
+    /*
+     * There might be concurrent writers to the keylog file, so we must ensure
+     * that one line is written at once.
+     */
+    BIO_puts(bio_keylog, line);
+    BIO_write(bio_keylog, "\n", 1);
+    (void)BIO_flush(bio_keylog);
+}
+
+int set_keylog_file(SSL_CTX *ctx, const char *keylog_file)
+{
+    /* Close any open files */
+    BIO_vfree(bio_keylog);
+    bio_keylog = NULL;
+
+    if (ctx == NULL || keylog_file == NULL) {
+        /* Keylogging is disabled, OK. */
+        return 0;
+    }
+
+    bio_keylog = BIO_new_file(keylog_file, "a");
+    if (bio_keylog == NULL) {
+        BIO_printf(bio_err, "Error writing keylog file %s\n", keylog_file);
+        return 1;
+    }
+
+    /* Write a header for seekable, empty files (this excludes pipes). */
+    if (BIO_tell(bio_keylog) == 0) {
```

Comment:
Append mode should be needed to prevent the file from overwriting previously-written secrets. This is the default behaviour for the SSLKEYLOGFILE environment variable and is just generally a useful idea, so append-mode is probably the way to go here.